src/java/org/apache/cassandra/service/accord/serializers/TopologySerializers.java

```diff
@@ -159,4 +167,127 @@ public long serializedSize(Topology topology)
             return size;
         }
     };
+
+    public static final UnversionedSerializer<Topology> compactTopology = new UnversionedSerializer<>()
+    {
+        @Override
+        public void serialize(Topology topology, DataOutputPlus out) throws IOException
+        {
+            out.writeLong(topology.epoch());
+            CollectionSerializers.serializeList(topology.staleIds(), out, TopologySerializers.nodeId);
+
+            List<Shard> shards = topology.shards();
+
+            // need to loop twice; once to collect tables/ranges, and another to save shards
+            ImmutableUniqueList<TableId> tables;
```

Comment:
I don't see why this class is needed for these serializers - why aren't we just using `Object2IntHashMap` directly? It would be more efficient, require no new types, and be more obvious to a reader. If you want to introduce a helper method for both serialising the array/list, and building the lookup that would make sense to me.

Reply:
> why aren't we just using Object2IntHashMap directly? 

Ill take a look at the issues that faces.

Reply:
ok i took a look again and its the following

1) `ImmutableUniqueList` is only *new* in the PR as git didn't see this as a migration; moved from `test` to `src`
2) serializing the list; I need to write the data in insertion order.  `Object2IntHashMap` isn't based off insertion order and isn't stable, so the serialization needs to do one of the following: copy the keys to a list and sorted by value, or maintain a `index -> type` mapping.  In both cases I argue `ImmutableUniqueList` is more efficient (the extra list doesn't need to be sorted, and is cheaper than the  `index -> type` mapping).

Reply:
Note, this logic works just as well for the case of duplicate items. You just skip anything already written, that must exist in the map. That is, you do

```
if (null == indexLookup.putIfAbsent(value, indexLookup.size()))
    out.write(value);
```

Reply:
> For deserialisation, the object->int mapping is unused, so this is a costly waste of effort to construct.

Agree, on deserialize it could be `List` directly

> The normal approach here is to simply write them in the order they come, and ensure they are inserted into the hash map in that same order, so that the hash map contains their index. This saves an array allocation.

We then loose the header saying how many items exist, but you try to address this in the last comment.

Ill take a look at option 1/2 and play around with this